File: Interview/Tymber/Genetic.java
import java.io.*;
import java.util.*;

public class Genetic {

    public static void main(String[] args) {
        String[] bank = {"GAAAAAAA", "AAGAAAAA", "AAAAGAAA", "GGAAAAAA", };

        int a = findMutationDistance("AAAAAAAA", "GGAAAAAA", bank);
        System.out.println("The distance is ");
        System.out.println(a);

    }

    public static int findMutationDistance(String start, String end, String[] bank) {

        Map<String, Integer> lpath = new HashMap<>();

        for (String word : bank) {
            lpath.put(word, -1);
        }

        if (!lpath.containsKey(end)) {
            return -1;
        }

        Queue<String> lqueue = new LinkedList<>();
        int counter = -1;
        lqueue.add(start);
        lpath.put(start, 0);

        while (!lqueue.isEmpty()) {
            String current = lqueue.poll();
            int pathSoFar = lpath.get(current);
            lpath.remove(current);
            if (current.equals(end)) {
                return pathSoFar;
            }
            List<String> next = nextWords(current, lpath);

            for (String word : next) {
                lpath.put(word, pathSoFar + 1);
                lqueue.add(word);
            }

        }
        return -1;
    }

    public static List<String> nextWords(String current, Map<String, Integer> lpath) {

        char[] tokens = {'A', 'C', 'T', 'G'};
        List<String> result = new LinkedList<>();
        for (int li = 0; li < current.length(); ++li) {
            char[] array = current.toCharArray();
            for (char c : tokens) {
                if (c == array[li]) { continue; }
                array[li] = c;
                String str = new String(array);
                if (lpath.containsKey(str)) { result.add(str); }
            }
        }
        return result;
    }

}

File: Interview/Twitter/genetic.py
# word -> GGAAAAAA
# c    -> A

def  findMutationDistance(start, end, bank):
    if end not in bank: return -1
    path_lengths = {word: -1 for word in bank}

    path_q = [] # pop(-1)
    path_q.append(start)
    path_lengths[start] = 0

    while any(path_q):
        current = path_q.pop(-1)
        path_length = path_lengths.pop(current)
        if current == end:
            return path_length

        next_words = nextWords(current, path_lengths)

        for word in next_words:
            path_lengths[word] = path_length + 1
            path_q.append(word)



def nextWords(current_word, path_lengths):
    tokens = {'A', 'C', 'T', 'G'}
    next_words = []

    for i in range(len(current_word)):
        for c in tokens:
            if c != current_word[i]:
                new_word = current_word[:i] + c + current_word[i+1:]
                if new_word in path_lengths: next_words.append(new_word)

    return next_words

if __name__ == '__main__':
    bank = {"GAAAAAAA", "AAGAAAAA", "AAAAGAAA", "GGAAAAAA"}
    print(findMutationDistance("AAAAAAAA", "GGAAAAAA", bank))
